samples/drivers/adc/src/main.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/printk.h>
+#include <drivers/adc.h>
+
+#if !DT_NODE_EXISTS(DT_PATH(zephyr_user)) || \
+	!DT_NODE_HAS_PROP(DT_PATH(zephyr_user), io_channels)
+#error "No suitable devicetree overlay specified"
+#endif
+
+#define ADC_NUM_CHANNELS	DT_PROP_LEN(DT_PATH(zephyr_user), io_channels)
+
+#if ADC_NUM_CHANNELS > 2
+#error "Currently only 1 or 2 channels supported in this sample"
+#endif
+
+/* Check that the same ADC is used for first and second phandle */
+#if ADC_NUM_CHANNELS == 2 && \
+	DT_REG_ADDR(DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 0)) \
+	!= DT_REG_ADDR(DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 1))
+#error "Channels have to use the same ADC."
+#endif
+
+#define ADC_NODE		DT_PHANDLE(DT_PATH(zephyr_user), io_channels)
+
+/* Common settings supported by most ADCs */
+#define ADC_RESOLUTION		12
+#define ADC_GAIN		ADC_GAIN_1
+#define ADC_REFERENCE		ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+
+/* Get the numbers of up to two channels */
+static uint8_t channel_ids[ADC_NUM_CHANNELS] = {
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 0),
+#if ADC_NUM_CHANNELS == 2
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 1)
+#endif
+};
+
+static int16_t sample_buffer[ADC_NUM_CHANNELS];
+
+struct adc_channel_cfg channel_cfg = {
+	.gain = ADC_GAIN,
+	.reference = ADC_REFERENCE,
+	.acquisition_time = ADC_ACQUISITION_TIME,
+	/* channel ID will be overwritten below */
+	.channel_id = 0,
+	.differential = 0
+};
+
+struct adc_sequence sequence = {
+	/* individual channels will be added below */
+	.channels    = 0,
+	.buffer      = sample_buffer,
+	/* buffer size in bytes, not number of samples */
+	.buffer_size = sizeof(sample_buffer),
+	.resolution  = ADC_RESOLUTION,
+};
+
+void main(void)
+{
+	int err;
+	int adc_vref, adc_value;
+	const struct device *dev_adc = DEVICE_DT_GET(ADC_NODE);
+
+	if (!device_is_ready(dev_adc)) {
+		printk("ADC device not found\n");
+		return;
+	}
+
+	/*
+	 * Configure channels individually prior to sampling
+	 */
+	for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
+		channel_cfg.channel_id = channel_ids[i];
```

Comment:
This patch makes this work on Nordic nRF52 devices:
```
@@ -78,6 +78,9 @@ void main(void)
 	 */
 	for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
 		channel_cfg.channel_id = channel_ids[i];
+#ifdef CONFIG_ADC_NRFX_SAADC
+		channel_cfg.input_positive = SAADC_CH_PSELP_PSELP_AnalogInput0 + channel_ids[i];
+#endif
 		adc_channel_setup(dev_adc, &channel_cfg);
 
 		sequence.channels |= BIT(channel_ids[i]);
```

Reply:
Ok, added your suggestion together with an overlay for nrf52dk_nrf52832. Unfortunately can't test because I don't have any Nordic board.

Reply:
It works for me:
```
*** Booting Zephyr OS build zephyr-v2.4.0-3073-g07bc2d3f55a5  ***               
ADC reading(s): 697 = 102 mV   771 = 112 mV                                     
ADC reading(s): 978 = 143 mV   4095 = 599 mV                                    
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): -415 = -61 mV   228 = 33 mV                                     
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): 4095 = 599 mV   4095 = 599 mV                                   
ADC reading(s): 4095 = 599 mV   4095 = 599 mV  
ADC reading(s): 1484 = 217 mV   1597 = 233 mV                                   
ADC reading(s): 992 = 145 mV   1121 = 164 mV   
```
(The max values are me putting my fingers on the AIN pins. Not sure what the negative is; perhaps the default configuration is double-ended.)